### ISSUE TITLE
Add support for FQDN Option from rfc 4704

### ISSUE THREAD

**Issue by traceflight** [bluecatengineering/dhcproto#58]
The FQDN option can be used to exchange information about a DHCPv6 client's Fully Qualified Domain Name (FQDN) and about responsibility for updating DNS resource records (RRs) related to the client's address assignments.
[RFC4704](https://www.rfc-editor.org/rfc/rfc4704)

**Comment by leshow** [comment]
I think there is some work left to do here for dhcpv6, we can leave this open if you want? You could manually pull the data out of `UnknownOption` but we support a first class type for v4 client fqdn, we could do the same for v6.

**Comment by stappersg** [comment]
On Thu, May 18, 2023 at 05:32:05PM -0700, Evan Cameron wrote:
> ... we support a first class type
> for v4 client fqdn, we could do the same for v6.

Any existing examples of such in v6 _and_ v4,
so this re-opened issue deserverse the label "good first issue"?


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by leshow** [comment]
https://github.com/bluecatengineering/dhcproto/blob/master/src/v4/fqdn.rs has an example for v4, I'd imagine v6 would work similarly and could share types if the flags section is the same (I think it is).

This is the RFC:
https://www.rfc-editor.org/rfc/rfc4704.html#section-4

should be pretty straightforward if anyone wants to take it.

**Comment by sage-s11** [comment]
I'd like to take the remaining v6 work.

Plan:
- Add a first-class `v6::DhcpOption::ClientFqdn` (option 39, RFC 4704 §4),
  same idea as `v4::fqdn::ClientFQDN` in `src/v4/fqdn.rs`.
- `OptionCode::ClientFqdn` / `OROCode::ClientFqdn` already exist; I'll wire
  encode/decode so this no longer lands in `UnknownOption`.

One difference from the note above: v4 and v6 flags are not the same, so I
won't share `v4::FqdnFlags`.
- v4 RFC 4702: `MBZ | N | E | O | S` plus RCODE1/RCODE2
- v6 RFC 4704: `MBZ | N | O | S` only (no E bit, no RCODE fields)
  S=0x01, O=0x02, N=0x04

I'll add a small `v6` flags + `ClientFqdn` type, DNS-wire domain name via
`hickory_proto::rr::Name` (canonical, no compression), and round-trip tests
for flags-only and a full FQDN. Partial names can stay out of the first PR
if `Name::read` is awkward there.

Happy to adjust if you'd rather share types anyway.

**Comment by leshow** [comment]
haven't looked at the RFC in a while but that seems like a good start. is there a reason not to use compression with the Name? does clientfqdn support compressed names?

**Comment by sage-s11** [comment]
I did give it a thought, but the RFC explicitly forbids compression (RFC 4704 §4.2 relies on RFC 8415 §10), so Client FQDN does not support compressed names. Because of this, I'm planning to emit/parse hickory_proto::rr::Name uncompressed, and keep the v6 flags separate from v4.

**Comment by leshow** [comment]
okay I had a little look,

On the flags, you're right that we cant use v4. I think we still need to copy v4's style though, set_n still implies N=1 & S=0.

I think we should handle partial and empty names in the first PR. If you only do one you turn what used to be an UnknownOption into a message parse failure for the other cases not handled.

Dont think Name::emit or set_canonical will work either, it writes the terminating zero regardless of is_fqdn, so partial and empty come out wrong on the wire. But you can still use Name. Double check but I think something like this is correct and handles the cases:

```
// encode
for label in name.iter() {
    e.write_u8(label.len())?;
    e.write_slice(label)?;
}
if name.is_fqdn() {
    e.write_u8(0)?;
}
```

```
// decode: labels until root or end
let mut name = Name::new();
loop {
    match dec.read_u8() {
        Ok(0) => { name.set_fqdn(true); break; }   // fqdn
        Ok(l) => name = name.append_label(dec.read_slice(l as usize)?)?,
        Err(_) => break,                            // partial name
    }
}
```

**Comment by sage-s11** [comment]
Agreed. Those three seem right - I was being optimistic about Name::emit.

set_n will match v4: N=1 forces S=0. Separate v6 flags type still, no E/RCODE.

Partial + empty have to ship in this PR. If decode only accepts a terminating root label we turn packets that used to land in UnknownOption into full message parse errors instead. RFC 4704 explicitly allows flags-only and partial names, and Name::read returns InsufficientBytes for both.

Confirmed on hickory-proto 0.26: emit_as_canonical always writes the final 0, is_fqdn is ignored. So we can keep Name for storage and do the label-loop encode/decode you sketched - 0 only when is_fqdn(), empty name via Name::new() not Name::root().

We can also reject any length byte >= 0xC0 on the way in, so a compression pointer doesn't get read as a 192-byte label.
